flagext/map_test.go

```diff
@@ -0,0 +1,431 @@
+				} else {
+					require.NoError(t, err)
+					require.Equal(t, tt.expected, lm.data)
+				}
+			})
+		}
+	})
+
+	t.Run("string", func(t *testing.T) {
+		tc := []struct {
+			name     string
+			input    string
+			expected map[string]string
+			error    string
+		}{
+			{
+				name: "unmarshal without error",
+				input: `
+key1: abc
+key2: def
+`,
+				expected: map[string]string{"key1": "abc", "key2": "def"},
+			},
+			{
+				name: "unmarshal with validation error",
+				input: `
+key1: abc
+key2: ""
+`,
+				error: "value cannot be empty",
+			},
+			{
+				name: "unmarshal with parsing error",
+				input: `
+key1: abc
+key2: def
+	key3: ghi
+`,
+				error: "yaml: line 3: found a tab character that violates indentation",
+			},
+		}
+
+		for _, tt := range tc {
+			t.Run(tt.name, func(t *testing.T) {
+				lm := NewLimitsMap(fakeStringValidator)
+				err := yaml.Unmarshal([]byte(tt.input), &lm)
+				if tt.error != "" {
+					require.Error(t, err)
+					require.Equal(t, tt.error, err.Error())
+				} else {
+					require.NoError(t, err)
+					require.Equal(t, tt.expected, lm.data)
+				}
+			})
+		}
+	})
+
+}
+
+func TestLimitsMap_MarshalYAML(t *testing.T) {
+	t.Run("numeric", func(t *testing.T) {
+		lm := NewLimitsMap(fakeFloat64Validator)
+		lm.data["key1"] = 10
+		lm.data["key2"] = 20
+
+		out, err := yaml.Marshal(&lm)
+		require.NoError(t, err)
+		require.Equal(t, "key1: 10\nkey2: 20\n", string(out))
+	})
+
+	t.Run("string", func(t *testing.T) {
+		lm := NewLimitsMap(fakeStringValidator)
+		lm.data["key1"] = "abc"
+		lm.data["key2"] = "def"
+
+		out, err := yaml.Marshal(&lm)
+		require.NoError(t, err)
+		require.Equal(t, "key1: abc\nkey2: def\n", string(out))
+	})
+}
+
+func TestLimitsMap_Equal(t *testing.T) {
+	t.Run("numeric", func(t *testing.T) {
+		tc := map[string]struct {
+			map1     LimitsMap[float64]
+			map2     LimitsMap[float64]
+			expected bool
+		}{
+			"Equal maps with same key-value pairs": {
+				map1:     LimitsMap[float64]{data: map[string]float64{"key1": 1.1, "key2": 2.2}},
+				map2:     LimitsMap[float64]{data: map[string]float64{"key1": 1.1, "key2": 2.2}},
+				expected: true,
+			},
+			"Different maps with different lengths": {
+				map1:     LimitsMap[float64]{data: map[string]float64{"key1": 1.1}},
+				map2:     LimitsMap[float64]{data: map[string]float64{"key1": 1.1, "key2": 2.2}},
+				expected: false,
+			},
+			"Different maps with same keys but different values": {
+				map1:     LimitsMap[float64]{data: map[string]float64{"key1": 1.1}},
+				map2:     LimitsMap[float64]{data: map[string]float64{"key1": 1.2}},
+				expected: false,
+			},
+			"Equal empty maps": {
+				map1:     LimitsMap[float64]{data: map[string]float64{}},
+				map2:     LimitsMap[float64]{data: map[string]float64{}},
+				expected: true,
+			},
+		}
+
+		for name, tt := range tc {
+			t.Run(name, func(t *testing.T) {
+				require.Equal(t, tt.expected, tt.map1.Equal(LimitsMap[float64]{data: tt.map2.data}))
```

Comment:
I'm not a huge fan of calling `require.Equal()` with booleans since the message if it fails is not helpful. Maybe split the tests into two groups and use the appropriate `require.Equal()` / `require.NotEqual()` on the values themselves?

Reply:
I've pushed a commit splitting them - however it's not quite as clean as I'd hoped.

Part of the reason it's written this way is to specifically test the `cmp.Equal` output and the custom equality function. `require.Equal()` internally only checks the same thing as the `==` operator, which doesn't necessarily map to the the Equal() function and can in theory differ.

Since `go-cmp` equality is a requirement, I still need a `require.True(t, cmp.Equal(...))` call as this test is trying to specifically cover the custom `Equal()` function.